src/librustc_typeck/check/demand.rs

```diff
@@ -58,7 +104,67 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         if let Err(e) = self.try_coerce(expr, checked_ty, expected) {
             let origin = TypeOrigin::Misc(expr.span);
             let expr_ty = self.resolve_type_vars_with_obligations(checked_ty);
-            self.report_mismatched_types(origin, expected, expr_ty, e);
+            let mode = probe::Mode::MethodCall;
+            let suggestions = if let Some(s) = self.check_ref(expr, checked_ty, expected) {
+                Some(s)
+            } else if let Ok(methods) = self.probe_return(syntax_pos::DUMMY_SP,
+                                                          mode,
+                                                          expected,
+                                                          checked_ty,
+                                                          ast::DUMMY_NODE_ID) {
+                let suggestions: Vec<_> =
+                    methods.iter()
+                           .map(|ref x| {
+                                Rc::new(x.item.clone())
+                            })
+                           .collect();
+                if suggestions.len() > 0 {
+                    Some(format!("here are some functions which \
+                                  might fulfill your needs:\n - {}",
```

Comment:
In similar cases, we didn't do it like this (the `#[rustc_on_unimplemented]` case is a good example). And I actually prefer without the help, even if the indent is ugly. Since we do not have something to do it yet, I think it's preferrable to keep it this way.